Failed to initialize NVML

The driver installed and loaded, distro rebooted. Any idea?
```
$ /usr/lib/wsl/lib/nvidia-smi
Failed to initialize NVML: GPU access blocked by the operating system
Failed to properly shut down NVML: GPU access blocked by the operating system
```
<img width="914" alt="Image" src="https://github.com/user-attachments/assets/89385aa5-4e25-493c-8120-eaf73a37b941" />

```
Building module:
Cleaning build area...
make -j12 KERNELRELEASE=6.12.12+bpo-amd64 -C /lib/modules/6.12.12+bpo-amd64/build M=/var/lib/dkms/dxgkrnl/be3283570/build...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.12.12+bpo-amd64 (x86_64)
Consult /var/lib/dkms/dxgkrnl/be3283570/build/make.log for more information.
Error! One or more modules failed to install during autoinstall.
Refer to previous errors for more information.
dkms: autoinstall for kernel: 6.12.12+bpo-amd64 failed!
run-parts: /etc/kernel/header_postinst.d/dkms exited with return code 11
Failed to process /etc/kernel/header_postinst.d at /var/lib/dpkg/info/linux-headers-6.12.12+bpo-amd64.postinst line 11.
dpkg: error processing package linux-headers-6.12.12+bpo-amd64 (--configure):
 installed linux-headers-6.12.12+bpo-amd64 package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 linux-headers-6.12.12+bpo-amd64
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

I'll try a 6.8 now, 6.12.12 doesn't seem to compile.

A lot of errors:

```
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_reclaim_allocations’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3621:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3621 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_change_vidmem_reservation’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3651:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3651 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_destroy_hwqueue’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3815:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3815 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_query_adapter_info’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3899:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3899 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_submit_command_hwqueue’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3957:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3957 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c: In function ‘dxgvmb_send_query_statistics’:
/var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3993:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
 3993 |                 DXG_TRACE("err: %d", ret);
      |                                          ^
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.o] Error 1
make[2]: *** Waiting for unfinished jobs....
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/hmgr.o] Error 1
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/dxgadapter.o] Error 1
cc1: some warnings being treated as errors
make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/ioctl.o] Error 1
make[1]: *** [/usr/src/linux-headers-6.12.12+bpo-common/Makefile:1962: /var/lib/dkms/dxgkrnl/be3283570/build] Error 2
make: *** [/usr/src/linux-headers-6.12.12+bpo-common/Makefile:236: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-headers-6.12.12+bpo-amd64'
```

Compiled a 6.9.10 though

> ```    
> /var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3993:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
>  3993 |                 DXG_TRACE("err: %d", ret);
>       |                                          ^
> cc1: some warnings being treated as errors
> make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.o] Error 1
> ```

```
cc1: some warnings being treated as errors
```

To ignore this, insert the ```-Wno-empty-body``` option into the compiler.

Here
install.sh:106:108
```
    # Patch a Makefile
    sed -i 's/\$(CONFIG_DXGKRNL)/m/' /usr/src/dxgkrnl-$VERSION/Makefile
    echo "EXTRA_CFLAGS=-I\$(PWD)/include -D_MAIN_KERNEL_" >> /usr/src/dxgkrnl-$VERSION/Makefile # !important
```

> > ```
> > /var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.c:3993:42: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
> >  3993 |                 DXG_TRACE("err: %d", ret);
> >       |                                          ^
> > cc1: some warnings being treated as errors
> > make[2]: *** [/usr/src/linux-headers-6.12.12+bpo-common/scripts/Makefile.build:234: /var/lib/dkms/dxgkrnl/be3283570/build/dxgvmbus.o] Error 1
> > ```
> 
> ```
> cc1: some warnings being treated as errors
> ```
> 
> To ignore this, insert the `-Wno-empty-body` option into the compiler.
> 
> Here install.sh:106:108
> 
> ```
>     # Patch a Makefile
>     sed -i 's/\$(CONFIG_DXGKRNL)/m/' /usr/src/dxgkrnl-$VERSION/Makefile
>     echo "EXTRA_CFLAGS=-I\$(PWD)/include -D_MAIN_KERNEL_" >> /usr/src/dxgkrnl-$VERSION/Makefile # !important
> ```

You are right about this, here is my `Makefile` that worked on `6.12.9`:

```
# SPDX-License-Identifier: GPL-2.0
# Makefile for the hyper-v compute device driver (dxgkrnl).

obj-m   += dxgkrnl.o
dxgkrnl-y := dxgmodule.o hmgr.o misc.o dxgadapter.o ioctl.o dxgvmbus.o dxgprocess.o  dxgsyncfile.o
EXTRA_CFLAGS=-I$(PWD)/include -D_MAIN_KERNEL_ -DCONFIG_DXGKRNL=m -include /usr/src/dxgkrnl-be3283570test/include/extra-defines.h -I /usr/src/linux-source-6.12/include/linux/ -include /usr/src/linux-source-6.12/include/linux/vmalloc.h -Wno-empty-body
```

--- a/install.sh
+++ b/install.sh
@@ -62,8 +62,10 @@ install() {
         "linux-msft-wsl-5.15.y")
             PATCHES="linux-msft-wsl-5.15.y/0001-Add-a-gpu-pv-support.patch \
                     linux-msft-wsl-5.15.y/0002-Add-a-multiple-kernel-version-support.patch";
-                    #linux-msft-wsl-5.15.y/0003-Fix-gpadl-has-incomplete-type-error.patch";
-
+            if [[ "$TARGET_KERNEL_VERSION" != *"azure"* ]]; then
+                    PATCHES="$PATCHES linux-msft-wsl-5.15.y/0003-Fix-gpadl-has-incomplete-type-error.patch";
+            fi
+            
             for PATCH in $PATCHES; do
                 # Patch source files
                 if [ -e "$WORKDIR/$PATCH" ]; then

--- a/install.sh
+++ b/install.sh
@@ -4,6 +4,9 @@ WORKDIR="$(dirname $(realpath $0))"
 LINUX_DISTRO="$(cat /etc/*-release)"
 LINUX_DISTRO=${LINUX_DISTRO,,}
 
+KERNEL_6_6_NEWER_REGEX="^(6\.[6-9]\.|6\.[0-9]{2,}\.)"
+KERNEL_5_15_NEWER_REGEX="^(5\.1[5-9]+\.)"
+
 install_dependencies() {
     if [[ "$LINUX_DISTRO" == *"debian"* ]]; then
         apt update;
@@ -27,11 +30,13 @@ install_dependencies() {
 }
 
 update_git() {
-    SYSTEM_KERNEL_VERSION="`echo ${TARGET_KERNEL_VERSION} | grep -Po ^[0-9]+\.[0-9]+`"
-    if [ "${SYSTEM_KERNEL_VERSION:0:1}" -ge "6" ] && [ "${SYSTEM_KERNEL_VERSION:2}" -ge "6" ]; then
+    if [[ "${TARGET_KERNEL_VERSION}" =~ $KERNEL_6_6_NEWER_REGEX ]]; then
         TARGET_BRANCH="linux-msft-wsl-6.6.y";
-    else
+    elif [[ "${TARGET_KERNEL_VERSION}" =~ $KERNEL_5_15_NEWER_REGEX ]]; then
         TARGET_BRANCH="linux-msft-wsl-5.15.y";
+    else
+        >&2 echo "Fatal: Unsupported kernel version (5.15.0 <=)";
+        exit 1;
     fi
 
     if [ ! -e "/tmp/WSL2-Linux-Kernel" ]; then

--- a/install.sh
+++ b/install.sh
@@ -113,7 +113,13 @@ install() {
     echo "EXTRA_CFLAGS=-I\$(PWD)/include -D_MAIN_KERNEL_ \
                        -I/usr/src/linux-headers-\${kernelver}/include/linux \
                        -include /usr/src/linux-headers-\${kernelver}/include/linux/vmalloc.h" >> /usr/src/dxgkrnl-$VERSION/Makefile # !important
-                       
+
+    if [[ "${TARGET_KERNEL_VERSION}" =~ $KERNEL_6_6_NEWER_REGEX ]]; then
+        BUILD_EXCLUSIVE_KERNEL=$KERNEL_6_6_NEWER_REGEX
+    else
+        BUILD_EXCLUSIVE_KERNEL=$KERNEL_5_15_NEWER_REGEX
+    fi
+
     # Create a config of DKMS
     # https://gist.github.com/krzys-h/e2def49966aa42bbd3316dfb794f4d6a
     cat > /usr/src/dxgkrnl-$VERSION/dkms.conf << EOF
@@ -122,6 +128,7 @@ PACKAGE_VERSION="$VERSION"
 BUILT_MODULE_NAME="dxgkrnl"
 DEST_MODULE_LOCATION="/kernel/drivers/hv/dxgkrnl/"
 AUTOINSTALL="yes"
+BUILD_EXCLUSIVE_KERNEL="$BUILD_EXCLUSIVE_KERNEL"
 EOF
 }
 

--- a/install.sh
+++ b/install.sh
@@ -110,8 +110,10 @@ install() {
 
     # Patch a Makefile
     sed -i 's/\$(CONFIG_DXGKRNL)/m/' /usr/src/dxgkrnl-$VERSION/Makefile
-    echo "EXTRA_CFLAGS=-I\$(PWD)/include -D_MAIN_KERNEL_" >> /usr/src/dxgkrnl-$VERSION/Makefile # !important
-
+    echo "EXTRA_CFLAGS=-I\$(PWD)/include -D_MAIN_KERNEL_ \
+                       -I/usr/src/linux-headers-\${kernelver}/include/linux \
+                       -include /usr/src/linux-headers-\${kernelver}/include/linux/vmalloc.h" >> /usr/src/dxgkrnl-$VERSION/Makefile # !important
+                       
     # Create a config of DKMS
     # https://gist.github.com/krzys-h/e2def49966aa42bbd3316dfb794f4d6a
     cat > /usr/src/dxgkrnl-$VERSION/dkms.conf << EOF

